### PR TITLE
build: use semantic-release to automate release

### LINKED ISSUE #1647: Add pull request template and policy
https://help.github.com/articles/about-issue-and-pull-request-templates/

Possible template/policies:
- PR title, which becomes the message of squashed merge commit
- Linking the issue with correct syntax (closing word)
- Adding an entry to CHANGELOG.md
- Updating `docs/supported.md`, `docs/support_table.md`, and other documentation if applicable
- Adding unit/regression/screenshot tests

### LINKED ISSUE #1780: Versioned docs are rendered using latest KaTeX
Currently, versioned docs are rendered using latest built KaTeX.

~This will be solved by Docusaurus 2, as it snapshots versioned docs.~

### LINKED ISSUE #2011: Use semantic-release to simplify releases
semantic-release determines what version to bump to when releasing based on commit messages, see https://github.com/semantic-release/semantic-release for details.  This may take a bit of work to integrate this into our existing release script.

As a first step though we could adopt https://github.com/semantic-release/commit-analyzer which would enforce a specific commit message style which would be helpful in manually determining which version to bump to.

### PR DESCRIPTION
**Pull request by ylemkimon**
Closes #2011. I'm really excited and surprised that it's really simple to use the `semantic-release`.

## Previous release process
* Run `release.sh` with the version to release
  * (1) Create a release branch
  * (2) Update `package.json` to the version to release
  * (3) Build
  * (4) Update `package.json` to the next version
  * (5) Update the README and documentation
  * (6) Generate a new version of the docs
  * (7) Make a release commit and push
* (8) Update CHANGELOG
* (9) Open PR
* (10) Update SRI hashes if files are changed
* Run `release.sh -p`
  * (11) Update `package.json` to the version to release
  * (12) Build distribution (`yarn dist`)
  * ~(13) Make another release commit~ **Removed in #2385**
  * (14) Make a tag and push
  * (15) Publish to `npm`
* (16) Create a GitHub release and upload artifacts

## New release process (fully automated)
* A PR is merged to a release branch, e.g., `master` ([Triggering a release](https://github.com/semantic-release/semantic-release#triggering-a-release))
* Determine the type of release based on commits added since the last release. ([@semantic-release/commit-analyzer](https://github.com/semantic-release/commit-analyzer))
  * [x] All PRs should follow [Angular Commit Message Conventions](https://github.com/angular/angular.js/blob/master/DEVELOPERS.md#-git-commit-guidelines)
  * [x] Updated CONTRIBUTING.md and added a pull request template (Fixes #1647)
* Generate release notes for commits added since the last release. ([@semantic-release/release-notes-generator](https://github.com/semantic-release/release-notes-generator))
* (8) Update CHANGELOG ([@semantic-release/changelog](https://github.com/semantic-release/changelog))
* (2, 11) Update `package.json` to the version to release ([@semantic-release/npm](https://github.com/semantic-release/npm))
  * `postversion`
    * (3, 12) Build distribution (`yarn dist`)
    * (5) Update the README and documentation
* (7) Make a release commit and push ([@semantic-release/git](https://github.com/semantic-release/git))
  * [x] Uses a bot account: @KaTeX-bot
  * [x] Permission to push to protected branch is required
  * [x] I've added branch protection rules to release branches.
* (14) Make a tag and push ([@semantic-release/git](https://github.com/semantic-release/git))
* (15) Publish to `npm` ([@semantic-release/npm](https://github.com/semantic-release/npm))
* (16) Create a GitHub release and upload artifacts ([@semantic-release/github](https://github.com/semantic-release/github))
* Notify of new releases or errors

### Removed
* (1) Creating a release branch and (9) opening a PR is no longer required
* There is no risk of SRI hashes changing and `semantic-release` is only run after all tests pass. So (10) updating hashes and running tests in (12) `yarn dist` is no longer required.

### Changed
* (6) Versioned docs are removed. They were broken anyway (#1780). Instead, links to previous Netlify deploys are used, https://?--katex.netlify.app/versions.html. Closes #1780.

### CODE REVIEW

**Comment by kevinbarabash** [comment]
@ylemkimon thanks for recreating this PR.  I'll review it this weekend.

File: .github/workflows/ci.yml
Comment (kevinbarabash) [inline .github/workflows/ci.yml:60]: Will this be uncommented in the future?  Should this be restricted to pushes to `master` like https://github.com/semantic-release/semantic-release/blob/master/docs/recipes/github-actions.md#githubworkflowsreleaseyml-configuration-for-node-projects?  When a push occurs, does the actual release only happen if the title of the commit is `Release: ...`?

File: website/siteConfig.js
Comment (kevinbarabash) [inline website/siteConfig.js:176]: Will the old versions on netlify be able to link back to the current version?

File: .releaserc
Comment (kevinbarabash) [inline .releaserc:7]: Once KaTeX reaches 1.0, these can be changed at that point, right?

File: package.json
Comment (kevinbarabash) [inline package.json:3]: Is the plan to only use real versions for release versions?  Should we also be changing the version back to this after a release is completed?

File: package.json
Comment (kevinbarabash) [inline package.json:107]: Does semantic release run this?

**Review by kevinbarabash — COMMENTED** [review]
This is awesome!  Thank you for setting this up.  I was excited for autogenerated release notes, but I didn't realize that this would totally do away with release.sh.  This makes things so much simpler and people don't need to learn about our custom release script.  I have a few questions inline about how parts of the process work, but this is looking very nice.

**Comment by kevinbarabash** [comment]
> A PR is merged to a release branch, e.g., master (Triggering a release)

> For each new commits added to one of the release branches (for example master, next, beta), with git push or by merging a pull request or merging from another branch, a CI build is triggered and runs the semantic-release command to make a release if there are codebase changes since the last release that affect the package functionalities.

Does this mean that a new release will occur whenever someone lands a PR with `feat: ...` in the title?

File: .github/workflows/ci.yml
Comment (ylemkimon) [reply .github/workflows/ci.yml:60]: > Will this be uncommented in the future?

Yes, I forgot to uncomment it. I just wanted to check it runs: https://github.com/KaTeX/KaTeX/runs/2089146698.

> Should this be restricted to pushes to `master` like https://github.com/semantic-release/semantic-release/blob/master/docs/recipes/github-actions.md#githubworkflowsreleaseyml-configuration-for-node-projects?

This already runs only on pushes to the master, but we may expand it to other branches for alpha/beta and backporting.

File: website/siteConfig.js
Comment (ylemkimon) [reply website/siteConfig.js:176]: Only versions after this PR, so it would work on versions after `v0.13.0`.

File: .releaserc
Comment (ylemkimon) [reply .releaserc:7]: Yes. In fact, the default option can be used.

File: package.json
Comment (ylemkimon) [reply package.json:3]: Yes, this will be only updated in released versions, which are uploaded to `npm`.

File: package.json
Comment (ylemkimon) [reply package.json:107]: Yes, after it updates the version but before packing the package.

**Comment by ylemkimon** [comment]
> Does this mean that a new release will occur whenever someone lands a PR with `feat: ...` in the title?

Yes, this way people wouldn't have to wait for a new release after a feature has landed.

**Comment by kevinbarabash** [comment]
Do bug fixes also trigger new releases?

**Comment by ylemkimon** [comment]
> Do bug fixes also trigger new releases?

Yes. Other scopes such as `chore:`, `doc:`, `build:`, `refactor:`, and `test:` don't trigger new releases.

**Review by kevinbarabash — APPROVED** [review]
Thanks for answering all my questions.  Publishing on every feat/bug is new to me.  If a change gets published that we have to revert for some reason is the plan to `deprecate` that package in npm?  There may be situations where it makes sense to do multiple PRs before releasing.  In that case, we could use a feature branch and then land that branch on to `master` (without squashing).  Situations like those should be rare though.  Thanks again for setting all of this up.  I'm looking forward to seeing it in action. 🚀

**Pull request merged**
